### PR TITLE
feature(adaptive-timeouts): calculate timeouts based on node load

### PR DESCRIPTION
**Pull request by soyacz**
During testing scenarios with different machine types and load it's
impossible to have one timeout value that fits all. Various operations
take more time when node performance is lower or data load is higher.

Implement `adaptive_timeout` context manager that will calculate right
timeout value for given operation based on node load (performance, data
size, and other factors). It is not raising any exceptions, just returns
timeout value that may be used within context. When context takes longer
duration than it, publish `SoftTimeoutEvent` with `Error` severity.

Because calculating timeout is not straightforward, collect various
'load metrics' to Elasticearch db for future reference and testing.

`SoftTimeoutEvent` has been refactored - dropped the context-manager
related code as it is no longer needed.

Supported `adaptive_timout` operations:
- `DECOMMISSION` - timeout for nodetool decommision operation.
Currently simple implementation, used for publishing `SoftTimeoutEvent`
- `SOFT_TIMEOUT` - passing timeout provided in args.

Other operations (like REPAIR, REBUILD, FLUSH etc. are created
only for collecting timeout data in ES for future reference when
designing timeout calcluators).

refs: https://github.com/scylladb/qa-tasks/issues/958

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by vponomaryov** [comment]
@soyacz do you have example test runs to look at?

**Comment by soyacz** [comment]
> @soyacz do you have example test runs to look at?

yes: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/lukasz/job/longevity-5gb-1h-DecommissionMonkey-aws-test/3/

**Comment by vponomaryov** [comment]
@soyacz 

In this test run I see following:

```
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR > 2023-03-10 15:12:32.960: (SoftTimeoutEvent Severity.ERROR) period_type=one-time event_id=e8670009-c79a-4f5f-a5d5-fc260e3ead89 during_nemesis=NodetoolDecommission, source=SoftTimeout message=operation 'DECOMMISSION' took 547.2067503929138s and soft-timeout was set to 188s
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR > Traceback (most recent call last):
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >   File "/usr/local/lib/python3.10/threading.py", line 966, in _bootstrap
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >     self._bootstrap_inner()
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >   File "/usr/local/lib/python3.10/threading.py", line 1009, in _bootstrap_inner
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >     self.run()
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >   File "/usr/local/lib/python3.10/threading.py", line 946, in run
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >     self._target(*self._args, **self._kwargs)
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >   File "/home/ubuntu/scylla-cluster-tests/sdcm/sct_events/decorators.py", line 26, in wrapper
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >     return func(*args, **kwargs)
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >   File "/home/ubuntu/scylla-cluster-tests/sdcm/nemesis.py", line 379, in run
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >     self.disrupt()
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >   File "/home/ubuntu/scylla-cluster-tests/sdcm/nemesis.py", line 4343, in disrupt
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >     self.disrupt_nodetool_decommission()
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >   File "/home/ubuntu/scylla-cluster-tests/sdcm/nemesis.py", line 4124, in wrapper
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >     result = method(*args[1:], **kwargs)
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >   File "/home/ubuntu/scylla-cluster-tests/sdcm/nemesis.py", line 1243, in disrupt_nodetool_decommission
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >     return self._nodetool_decommission(add_node=add_node)
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >   File "/home/ubuntu/scylla-cluster-tests/sdcm/nemesis.py", line 1221, in _nodetool_decommission
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >     self.cluster.decommission(self.target_node)
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >   File "/home/ubuntu/scylla-cluster-tests/sdcm/cluster.py", line 4532, in decommission
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >     with adaptive_timeout(operation=Operations.DECOMMISSION, node=node):
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >   File "/usr/local/lib/python3.10/contextlib.py", line 142, in __exit__
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >     next(self.gen)
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >   File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/adaptive_timeouts/__init__.py", line 51, in adaptive_timeout
17:12:39  < t:2023-03-10 15:12:32,976 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:ERROR >     SoftTimeoutEvent(operation=operation.name, soft_timeout=timeout, duration=duration).publish_or_dump()
```

And it is not how I expect to see the `SCT event` to happen. I don't expect a traceback in this case.
Or it is to be fixed? Or am I missing something?

**Comment by soyacz** [comment]
> 
> And it is not how I expect to see the `SCT event` to happen. I don't expect a traceback in this case. Or it is to be fixed? Or am I missing something?

I added traceback to keep the same behavior like it was with SoftTimeout context manager I removed. I'm also not sure if we need it - to be decided.

File: sdcm/utils/adaptive_timeouts/load_info_store.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+        # pylint: disable=unexpected-keyword-arg,no-value-for-parameter
+        self._es.create(index=self._index, id=load_id, document=body)
+
+    def get(self, operation: str | None, timeout_occurred: bool | None = None):
+        """Get adaptive timeout info from ES.
+
+        Example usage:
+            >>> from sdcm.utils.adaptive_timeouts.load_info_store import ESAdaptiveTimeoutStore
+            >>> from sdcm.utils.adaptive_timeouts import Operations
+            >>> ESAdaptiveTimeoutStore().get(operation=Operations.DECOMMISSION.name)
```
Comment (soyacz) [inline sdcm/utils/adaptive_timeouts/load_info_store.py:182]: I updated some code to view metrics/data from ES with example. I already fixed missing version details. 

Also looks like this calculation needs some tuning - when CPU usage is high it soft-times-out

File: sdcm/cluster_k8s/__init__.py [hunk 2]
```diff
@@ -2670,9 +2670,7 @@ def decommission(self, node: BaseScyllaPodContainer, timeout: int | float = None
         # node deletion using "terminate_node" command.
         scylla_shards = node.scylla_shards
 
-        # NOTE: on k8s we treat soft_timeout as the "hard" timeout,
-        # at least until we'll have case with bigger data sets in it
-        timeout = timeout or soft_timeout or (node.pod_terminate_timeout * 60)
+        timeout = timeout or (node.pod_terminate_timeout * 60)
```
Comment (fruch) [inline sdcm/cluster_k8s/__init__.py:2673]: Try to add the `adaptive_timeout` also in k8s case

**Comment by soyacz** [comment]
> 
> And it is not how I expect to see the `SCT event` to happen. I don't expect a traceback in this case. Or it is to be fixed? Or am I missing something?

It mimic's the logic of refactored `SoftTimeoutEvent` - traceback was also there.

File: sdcm/cluster_k8s/__init__.py (hunk 2 above)
Comment (soyacz) [reply sdcm/cluster_k8s/__init__.py:2674]: done

**Comment by soyacz** [comment]
test drive on EKS: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/lukasz/job/longevity-scylla-operator-3h-eks/3/

**Comment by soyacz** [comment]
test drive failed on running `nodetool info` command:
```
invoke.exceptions.UnexpectedExit: Encountered a bad command exit code!
Command: 'nodetool info'
Exit code: 1
Stdout:
Off Heap Memory (MB)   : 338.42
Data Center            : us-east1-b
Rack                   : us-east1
Exceptions             : 0
Key Cache              : entries 0, size 0 bytes, capacity 0 bytes, 0 hits, 0 requests, 0.000 recent hit rate, 0 save period in seconds
Row Cache              : entries 17, size 17 bytes, capacity 808.32 KiB, 630 hits, 1400 requests, 0.450 recent hit rate, 0 save period in seconds
Counter Cache          : entries 0, size 0 bytes, capacity 0 bytes, 0 hits, 0 requests, 0.000 recent hit rate, 0 save period in seconds
Percent Repaired       : 0.0%
nodetool: Scylla API server HTTP GET to URL '/storage_service/tokens/null' failed: std::invalid_argument (null)
See 'nodetool help' or 'nodetool help <command>'.
```
@vponomaryov do you know the issue?

**Comment by fruch** [comment]
Sounds like we aren not the first to hit it:
https://github.com/scylladb/scylla-jmx/issues/59

This is called after decommission has stated/ended ?

**Comment by fgelcer** [comment]
> Sounds like we aren not the first to hit it: [scylladb/scylla-jmx#59](https://github.com/scylladb/scylla-jmx/issues/59)
> 
> This is called after decommission has stated/ended ?

this is an issue from 2018...

**Comment by fruch** [comment]
> > Sounds like we aren not the first to hit it: [scylladb/scylla-jmx#59](https://github.com/scylladb/scylla-jmx/issues/59)
> > This is called after decommission has stated/ended ?
> 
> this is an issue from 2018...

it doesn't mean someone fixed it :)

it's identical to what @soyacz is doing here, I think calling `nodetool info` might if used after decommission (and probably lots of other `nodetool` commands)

this would happen if host_id is null, or failed to retrieve it
https://github.com/scylladb/scylla-jmx/blob/88d9bdc5b28ac9fb7a4d36f59aa47ea2c5046659/src/main/java/org/apache/cassandra/service/StorageService.java#LL423-L427C6

**Comment by soyacz** [comment]
> > > Sounds like we aren not the first to hit it: [scylladb/scylla-jmx#59](https://github.com/scylladb/scylla-jmx/issues/59)
> > > This is called after decommission has stated/ended ?
> > 
> > 
> > this is an issue from 2018...
> 
> it doesn't mean someone fixed it :)
> 
> it's identical to what @soyacz is doing here, I think calling `nodetool info` might if used after decommission (and probably lots of other `nodetool` commands)
> 
> this would happen if host_id is null, or failed to retrieve it https://github.com/scylladb/scylla-jmx/blob/88d9bdc5b28ac9fb7a4d36f59aa47ea2c5046659/src/main/java/org/apache/cassandra/service/StorageService.java#LL423-L427C6

Yes, I placed `adaptive_timeout` context manager after triggering decommission, only on wait function (just to be more precise about decommission time only...). I'll place it before triggering decommission.

**Comment by soyacz** [comment]
Some problems were found with getting scylla metrics, fixed, retry: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/lukasz/job/longevity-scylla-operator-3h-eks/10/

**Review by fruch — APPROVED** [review]
LGTM

**Review by fruch — DISMISSED** [review]
LGTM

**Comment by roydahan** [comment]
ok.

**Review by fruch — APPROVED** [review]

**Pull request merged**